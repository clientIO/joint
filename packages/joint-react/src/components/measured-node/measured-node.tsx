import { forwardRef, memo } from 'react';
import { useChildrenRef } from '../../hooks/use-children-ref';
import { useMeasureNodeSize, type MeasureNodeOptions } from '../../hooks/use-measure-node-size';

export interface MeasuredNodeProps extends MeasureNodeOptions {
  /**
   * The child element to measure.
   * It can be only HTML or SVG element.
   */
  readonly children: React.ReactNode | null;
}

// eslint-disable-next-line jsdoc/require-jsdoc
function Component(
  props: MeasuredNodeProps,
  forwardedRef: React.ForwardedRef<HTMLElement | SVGAElement>
) {
  const { children, ...options } = props;
  const { elementRef, elementChildren } = useChildrenRef(children, forwardedRef);
  useMeasureNodeSize(elementRef, options);
  return elementChildren;
}

const ForwardedRefComponent = forwardRef(Component);

/**
 * Measured node component automatically detects the size of its `children` and updates the graph element (node) width and height automatically when elements resize.
 *
 * It must be used inside `renderElement` context
 * @see Paper
 * @see PaperProps
 * @group Components
 * @example
 * Example with a simple div:
 * ```tsx
 * import { MeasuredNode } from '@joint/react';
 *
 * function RenderElement() {
 *   return (
 *     <MeasuredNode>
 *       <div style={{ width: 100, height: 50 }}>Content</div>
 *     </MeasuredNode>
 *   );
 * }
 * ```
 *
 * Example with a simple div without explicit size defined:
 * ```tsx
 * import { MeasuredNode } from '@joint/react';
 *
 * function RenderElement() {
 *   return (
 *     <MeasuredNode>
 *       <div style={{ padding: 10 }}>Content</div>
 *     </MeasuredNode>
 *   );
 * }
 * ```
 * @example
 * Example with custom size handling:
 * ```tsx
 * import { MeasuredNode } from '@joint/react';
 * import type { dia } from '@joint/core';
 *
 * function RenderElement() {
 *   const handleSizeChange = (element: dia.Cell, size: { width: number; height: number }) => {
 *     console.log('New size:', size);
 *     element.set('size', { width: size.width + 10, height: size.height + 10 });
 *   };
 *
 *   return (
 *     <MeasuredNode setSize={handleSizeChange}>
 *       <div style={{ width: 100, height: 50 }}>Content</div>
 *     </MeasuredNode>
 *   );
 * }
 * ```
 */
export const MeasuredNode = memo(ForwardedRefComponent);
