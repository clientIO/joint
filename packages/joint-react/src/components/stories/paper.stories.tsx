/* eslint-disable sonarjs/no-small-switch */

/* eslint-disable no-console */
/* eslint-disable react-perf/jsx-no-new-object-as-prop */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */

import type { Meta, StoryObj } from '@storybook/react'
import type { dia } from '@joint/core'

import { shapes } from '@joint/core'
import { GraphProvider } from '../graph-provider'
import { PaperProvider } from '../paper-provider'
import { Paper } from '../paper'
import { useGraphStore } from '../../hooks/use-graph-store'
import { useGraphCells } from '../../hooks/use-graph-cells'
import { CellsExplorer } from './cell-explorer'
import { ReactElement } from '../../models/react-element'

import { useSetGraphCells } from '../../hooks/use-set-graph-cells'

const paperStoryOptions: dia.Paper.Options = {
  width: 400,
  height: 400,
  background: { color: '#f8f9fa' },
  gridSize: 2,
}

// elements to add to the graph
const ELEMENTS = () => {
  return [
    new shapes.standard.Rectangle({
      id: '1',
      position: { x: 100, y: 30 },
      size: { width: 100, height: 40 },
      attrs: {
        label: { text: 'rect-1' },
        body: { fill: 'blue' },
      },
    }),
    new shapes.standard.Rectangle({
      id: '2',
      position: { x: 100, y: 100 },
      size: { width: 100, height: 40 },
      attrs: {
        label: { text: 'rect-2' },
        body: { fill: 'red' },
      },
    }),
    new shapes.standard.Link({
      source: { id: '1' },
      target: { id: '2' },
      attrs: {
        line: { stroke: 'blue', targetMarker: { name: 'classic', size: 8 } },
      },
    }),
    new ReactElement({
      id: '3',
      position: { x: 100, y: 200 },
      size: { width: 100, height: 40 },
      attrs: {
        body: { fill: 'green' },
      },
    }),

    new ReactElement({
      id: '4',
      position: { x: 100, y: 300 },
      size: { width: 100, height: 40 },
      attrs: {
        body: { fill: 'green' },
      },
    }),
  ]
}
function UpdateCellsViaGraphApi() {
  const { graph } = useGraphStore()

  return (
    <div>
      <div>Updating cells via graph API</div>
      <button onClick={() => graph.clear()}>Remove all cells</button>
      <button onClick={() => graph.addCells(ELEMENTS())}>Add all cells</button>
    </div>
  )
}

const meta: Meta<typeof Paper> = {
  title: 'Components/Paper',
  component: Paper,
}

export default meta
export type PaperStory = StoryObj<typeof Paper>

function CellsExplorerViaHook() {
  const cells = useGraphCells((cell) => cell.toJSON())
  const setCells = useSetGraphCells()
  return <CellsExplorer cells={cells} onChange={setCells} />
}

export const WithHooksAPI: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider>
        <UpdateCellsViaGraphApi />
        <div style={{ display: 'flex', flex: 1 }}>
          <PaperProvider {...paperStoryOptions}>
            <Paper
              elementSelector={(cell) => {
                switch (cell instanceof ReactElement) {
                  case true: {
                    return { id: cell.id, x: cell.attributes.position?.x }
                  }
                  default: {
                    return {
                      id: cell.id,
                    }
                  }
                }
              }}
              renderElement={(cell) => (
                <div style={{ width: '100%', height: '10%' }} onClick={() => console.log('CLICK')}>
                  x: {cell.x}
                </div>
              )}
            />
          </PaperProvider>
          <CellsExplorerViaHook />
        </div>
      </GraphProvider>
    )
  },
}

export { PaperStressTestNative, PaperStressTestReact } from './paper-stress'
