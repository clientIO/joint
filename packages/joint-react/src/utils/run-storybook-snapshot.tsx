/* eslint-disable unicorn/consistent-function-scoping */
/* eslint-disable jsdoc/require-jsdoc */
import { render } from '@testing-library/react';
import { GraphProvider, Paper } from '../components';

interface Options<StorybookOptions> {
  Component: React.ComponentType<StorybookOptions>;
  name: string;
  // we have to use any here because we don't know the type of the story
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  stories: Record<string, any>;
  withRenderElementWrapper?: boolean;
}

function normalizeSnapshot(html: string) {
  const cleaned = html.replaceAll(/id="v-\d+"/g, 'id="v-*"');

  return cleaned;
}

/**
 * Runs a snapshot test for each story in the provided options.
 * @param options - The options for the snapshot test.
 * @param options.Component - The component to render.
 * @param options.name - The name of the test suite.
 * @param options.stories - The stories to test.
 */
export function runStorybookSnapshot<StorybookOptions>(options: Options<StorybookOptions>) {
  const { stories, Component, name, withRenderElementWrapper } = options;
  const keys = Object.keys(stories).filter((key) => key !== 'default');
  describe(name, () => {
    it.each(keys)('%p', (key) => {
      const story = stories[key as keyof typeof stories];

      const props = ('args' in story ? story.args : {}) as StorybookOptions & {
        children?: React.ReactNode;
      };

      let wrapper = function ({ children }: { children: React.ReactNode }) {
        return <>{children}</>;
      };
      if (withRenderElementWrapper) {
        wrapper = function ({ children }: { children: React.ReactNode }) {
          return (
            <GraphProvider>
              <Paper renderElement={children as never} />
            </GraphProvider>
          );
        };
      }
      // ACT
      const { container } = render(<Component {...props} />, { wrapper });

      // Clean volatile parts before asserting
      const cleanedHTML = normalizeSnapshot(container.innerHTML);
      expect(cleanedHTML).toMatchSnapshot(`${name}-${key}`);
    });
  });
}
