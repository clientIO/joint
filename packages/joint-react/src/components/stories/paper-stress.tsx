/* eslint-disable react-perf/jsx-no-new-object-as-prop */
/* eslint-disable sonarjs/pseudo-random */
/* eslint-disable no-console */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */
import { dia, shapes } from '@joint/core'
import { ReactElement } from '../../models/react-element'
import { GraphProvider } from '../graph-provider'
import { Paper } from '../paper'
import { PaperProvider } from '../paper-provider'
import type { PaperStory } from './paper.stories'
import { useSetCells } from '../../hooks/use-set-cells'
import { updateGraph } from '../../utils/update-graph'
import { useElements } from '../../hooks/use-elements'

const paperStoryOptions: dia.Paper.Options = {
  width: 400,
  height: 400,
  background: { color: '#f8f9fa' },
  gridSize: 2,
}
const graph = new dia.Graph({}, { cellNamespace: { ...shapes, ReactElement } })
const notUsedGraph = new dia.Graph({}, { cellNamespace: { ...shapes, ReactElement } })
function createElements(xCount: number, yCount: number) {
  const elements = []
  const ELEMENT_SIZE = 50
  const MARGIN = 2
  for (let x = 0; x < xCount; x++) {
    for (let y = 0; y < yCount; y++) {
      elements.push(
        new ReactElement({
          id: `${x}-${y}`,
          position: { x: x * (ELEMENT_SIZE + MARGIN), y: y * (ELEMENT_SIZE + MARGIN) },
          size: { width: ELEMENT_SIZE, height: ELEMENT_SIZE },
          attrs: {
            label: { text: `${x}-${y}` },
            body: { fill: 'blue', stroke: 'black' },
          },
        })
      )
      // add a link to the previous element
      if (x > 0) {
        elements.push(
          new shapes.standard.Link({
            id: `${x - 1}-${y}-${x}-${y}`,
            source: { id: `${x - 1}-${y}` },
            target: { id: `${x}-${y}` },
          })
        )
      }
    }
  }
  return elements
}

graph.addCells(createElements(15, 30))
notUsedGraph.addCells(createElements(15, 30))
function RandomChange() {
  const elementsSize = useElements((items) => items.map((item) => item.id))
  const setCells = useSetCells()
  console.log('re-render RandomChange Component', elementsSize)
  return (
    <>
      <button
        onClick={() => {
          console.time('Random move')
          setCells((previousCells) =>
            previousCells.map((cell) => {
              if (cell instanceof ReactElement) {
                cell.set({
                  position: { x: Math.random() * 1000, y: Math.random() * 1000 },
                })
                return cell
              }
              return cell
            })
          )
          console.timeEnd('Random move')
        }}
      >
        {/* Random move {elementsSize.length} elements */}
      </button>

      <button
        onClick={() => {
          console.time('Random move with notUsedGraph')
          const oldCells = notUsedGraph.getCells()
          updateGraph(
            notUsedGraph,
            oldCells.map((cell) => {
              if (cell instanceof ReactElement) {
                cell.set({
                  position: { x: Math.random() * 1000, y: Math.random() * 1000 },
                })
                return cell
              }
              return cell
            })
          )
          console.timeEnd('Random move with notUsedGraph')
        }}
      >
        Random move with notUsedGraph - only prints console.logs
      </button>
    </>
  )
}

export const PaperStressTestNative: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          <PaperProvider {...paperStoryOptions}>
            <Paper />
          </PaperProvider>
        </div>
      </GraphProvider>
    )
  },
}

export const PaperStressTestReact: PaperStory = {
  args: {
    style: { border: '1px solid #ccc' },
  },
  render: () => {
    console.log('re-render WithHooksAPI')
    return (
      <GraphProvider graph={graph}>
        <RandomChange />
        <div style={{ display: 'flex', flex: 1 }}>
          {/* <PaperProvider {...paperStoryOptions}> */}
          <Paper
            {...paperStoryOptions}
            // elementSelector={(cell) => {
            //   switch (cell instanceof ReactElement) {
            //     case true: {
            //       return { id: cell.id, xPosition: cell.attributes.position?.x }
            //     }
            //     default: {
            //       return {
            //         id: cell.id,
            //       }
            //     }
            //   }
            // }}
            renderElement={(element) => {
              console.log('re-render renderElement')
              return (
                <div style={{ fontSize: 12 }} onClick={() => console.log('Click from React')}>
                  {JSON.stringify(element.attributes.position)}
                </div>
              )
            }}
          />
          {/* </PaperProvider> */}
        </div>
      </GraphProvider>
    )
  },
}
