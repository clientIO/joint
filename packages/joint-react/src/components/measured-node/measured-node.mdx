import { Meta, Story, Canvas, Controls, Markdown } from '@storybook/blocks';
import * as Stories from './measured-node.stories';

<Meta of={Stories} />

# MeasuredNode Component
Measured node component automatically detects the size of its `children` and 
updates the graph element (`node`) **width and height** automatically when elements resize.
It must be used inside `renderElement` context - inside `Paper` component.

### Measuring div with exact size defined in the div style.
<Canvas of={Stories.DivWithExactSize} />

### Measuring div with padding and text inside (without explicit size).
<Canvas of={Stories.DivWithPaddingAndText} />

It support also `onSetSize` callback that can be used to update the size of the node (it turn off default functionality to change the size of the node).
<Controls/>
