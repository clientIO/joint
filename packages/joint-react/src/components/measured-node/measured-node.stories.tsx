/* eslint-disable react-perf/jsx-no-new-object-as-prop */

import type { Meta, StoryObj } from '@storybook/react/*';
import { SimpleRenderItemDecorator } from '../../../.storybook/decorators/with-simple-data';
import { MeasuredNode } from './measured-node';
import { useElement } from '@joint/react/src/hooks/use-element';
import { PRIMARY } from 'storybook-config/theme';
import { makeRootDocumentation, makeStory } from '@joint/react/src/stories/utils/make-story';
import { getAPILink } from '@joint/react/src/stories/utils/get-api-documentation-link';

const API_URL = getAPILink('MeasuredNode', 'variables');

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function ForeignObjectDecorator(Story: any) {
  const { width, height } = useElement();
  return (
    <foreignObject overflow="visible" width={width} height={height}>
      <Story />
    </foreignObject>
  );
}
export type Story = StoryObj<typeof MeasuredNode>;

const meta: Meta<typeof MeasuredNode> = {
  title: 'Components/MeasuredNode',
  component: MeasuredNode,
  decorators: [ForeignObjectDecorator, SimpleRenderItemDecorator],
  parameters: makeRootDocumentation({
    apiURL: API_URL,
    code: `import { MeasuredNode } from '@joint/react'
// This will automatically measure component size and update the parent node size
<MeasuredNode>
  <div style={{ width: 100, height: 50 }}>Content</div>
</MeasuredNode>
    `,
    description: `
Measured node component automatically detects the size of its \`children\` and updates the graph element (node) width and height automatically when elements resize.
It must be used inside \`renderElement\` context. 
    `,
  }),
};

export default meta;

export const DivWithExactSize = makeStory<Story>({
  args: {
    children: (
      <div style={{ width: 100, height: 50, backgroundColor: PRIMARY, borderRadius: 10 }} />
    ),
  },
  apiURL: API_URL,
  name: 'Measured div with exact size',
  description: 'Div with exact size.',
});

export const DivWithPaddingAndText = makeStory<Story>({
  args: {
    children: (
      <div
        style={{
          padding: 10,
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          textAlign: 'center',
          backgroundColor: PRIMARY,
          borderRadius: 10,
        }}
      >
        Hello world!
      </div>
    ),
  },
  apiURL: API_URL,
  name: 'Measured div with padding and text',
  description: 'Div with padding and text content.',
});

export const TailwindSizing = makeStory<Story>({
  args: {
    children: (
      <div className="flex items-center justify-center text-center bg-primary rounded-lg p-2 bg-red-500">
        Hello world!
      </div>
    ),
  },
  apiURL: API_URL,
  name: 'Tailwind sizing',
  description: 'Div with tailwind classes.',
});
