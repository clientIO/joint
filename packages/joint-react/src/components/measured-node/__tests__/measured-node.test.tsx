import { runStorybookSnapshot } from '../../../utils/run-storybook-snapshot';
import { MeasuredNode } from '../measured-node';
import * as stories from '../measured-node.stories';
runStorybookSnapshot({
  Component: MeasuredNode,
  stories,
  name: 'MeasuredNode',
  withRenderElementWrapper: true,
});
