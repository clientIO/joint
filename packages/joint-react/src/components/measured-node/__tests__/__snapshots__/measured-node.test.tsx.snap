// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasuredNode "DivWithExactSize" 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
      >
        <div
          class="joint-paper joint-theme-default"
          style="position: relative; width: 800px; height: 600px;"
        >
          <div
            class="joint-paper-background"
            joint-selector="background"
            style="position: absolute; inset: 0;"
          />
          <svg
            height="100%"
            id="v-2"
            joint-selector="svg"
            style="position: absolute; inset: 0; overflow: hidden;"
            width="100%"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <style
              id="v-3"
              type="text/css"
            >
              CDATASection {}
            </style>
            <defs
              joint-selector="defs"
            />
            <g
              class="joint-layers"
              joint-selector="layers"
            >
              <g
                class="joint-grid-layer"
                style="pointer-events: none;"
              />
              <g
                class="joint-back-layer"
              />
              <g
                class="joint-cells-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-labels-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-front-layer"
              />
              <g
                class="joint-tools-layer"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="joint-paper joint-theme-default"
        style="position: relative; width: 800px; height: 600px;"
      >
        <div
          class="joint-paper-background"
          joint-selector="background"
          style="position: absolute; inset: 0;"
        />
        <svg
          height="100%"
          id="v-2"
          joint-selector="svg"
          style="position: absolute; inset: 0; overflow: hidden;"
          width="100%"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <style
            id="v-3"
            type="text/css"
          >
            CDATASection {}
          </style>
          <defs
            joint-selector="defs"
          />
          <g
            class="joint-layers"
            joint-selector="layers"
          >
            <g
              class="joint-grid-layer"
              style="pointer-events: none;"
            />
            <g
              class="joint-back-layer"
            />
            <g
              class="joint-cells-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-labels-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-front-layer"
            />
            <g
              class="joint-tools-layer"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MeasuredNode "DivWithPaddingAndText" 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
      >
        <div
          class="joint-paper joint-theme-default"
          style="position: relative; width: 800px; height: 600px;"
        >
          <div
            class="joint-paper-background"
            joint-selector="background"
            style="position: absolute; inset: 0;"
          />
          <svg
            height="100%"
            id="v-4"
            joint-selector="svg"
            style="position: absolute; inset: 0; overflow: hidden;"
            width="100%"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <style
              id="v-5"
              type="text/css"
            >
              CDATASection {}
            </style>
            <defs
              joint-selector="defs"
            />
            <g
              class="joint-layers"
              joint-selector="layers"
            >
              <g
                class="joint-grid-layer"
                style="pointer-events: none;"
              />
              <g
                class="joint-back-layer"
              />
              <g
                class="joint-cells-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-labels-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-front-layer"
              />
              <g
                class="joint-tools-layer"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="joint-paper joint-theme-default"
        style="position: relative; width: 800px; height: 600px;"
      >
        <div
          class="joint-paper-background"
          joint-selector="background"
          style="position: absolute; inset: 0;"
        />
        <svg
          height="100%"
          id="v-4"
          joint-selector="svg"
          style="position: absolute; inset: 0; overflow: hidden;"
          width="100%"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <style
            id="v-5"
            type="text/css"
          >
            CDATASection {}
          </style>
          <defs
            joint-selector="defs"
          />
          <g
            class="joint-layers"
            joint-selector="layers"
          >
            <g
              class="joint-grid-layer"
              style="pointer-events: none;"
            />
            <g
              class="joint-back-layer"
            />
            <g
              class="joint-cells-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-labels-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-front-layer"
            />
            <g
              class="joint-tools-layer"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MeasuredNode "TailwindSizing" 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
      >
        <div
          class="joint-paper joint-theme-default"
          style="position: relative; width: 800px; height: 600px;"
        >
          <div
            class="joint-paper-background"
            joint-selector="background"
            style="position: absolute; inset: 0;"
          />
          <svg
            height="100%"
            id="v-6"
            joint-selector="svg"
            style="position: absolute; inset: 0; overflow: hidden;"
            width="100%"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <style
              id="v-7"
              type="text/css"
            >
              CDATASection {}
            </style>
            <defs
              joint-selector="defs"
            />
            <g
              class="joint-layers"
              joint-selector="layers"
            >
              <g
                class="joint-grid-layer"
                style="pointer-events: none;"
              />
              <g
                class="joint-back-layer"
              />
              <g
                class="joint-cells-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-labels-layer joint-viewport"
                style="webkit-user-select: none; user-select: none;"
              />
              <g
                class="joint-front-layer"
              />
              <g
                class="joint-tools-layer"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="joint-paper joint-theme-default"
        style="position: relative; width: 800px; height: 600px;"
      >
        <div
          class="joint-paper-background"
          joint-selector="background"
          style="position: absolute; inset: 0;"
        />
        <svg
          height="100%"
          id="v-6"
          joint-selector="svg"
          style="position: absolute; inset: 0; overflow: hidden;"
          width="100%"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <style
            id="v-7"
            type="text/css"
          >
            CDATASection {}
          </style>
          <defs
            joint-selector="defs"
          />
          <g
            class="joint-layers"
            joint-selector="layers"
          >
            <g
              class="joint-grid-layer"
              style="pointer-events: none;"
            />
            <g
              class="joint-back-layer"
            />
            <g
              class="joint-cells-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-labels-layer joint-viewport"
              style="webkit-user-select: none; user-select: none;"
            />
            <g
              class="joint-front-layer"
            />
            <g
              class="joint-tools-layer"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
